File: backend/app/api/projects.py
from fastapi import APIRouter, HTTPException, Depends, status
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from app.services.project_service import project_service
from app.services.cache_service import cache_service
from app.api.auth import get_current_user

router = APIRouter(prefix="/projects", tags=["projects"])

class ProjectCreate(BaseModel):
    name: str
    description: Optional[str] = None
    project_type: str = "web"

class ProjectUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    settings: Optional[Dict[str, Any]] = None

class ProjectResponse(BaseModel):
    id: str
    user_id: str
    name: str
    description: Optional[str]
    project_type: str
    settings: Dict[str, Any]
    is_active: bool
    created_at: str
    updated_at: str

@router.post("/", response_model=ProjectResponse)
async def create_project(
    project_data: ProjectCreate,
    current_user: dict = Depends(get_current_user)
):
    """Create a new project"""
    try:
        project = await project_service.create_project(
            user_id=current_user["id"],
            name=project_data.name,
            description=project_data.description,
            project_type=project_data.project_type
        )
        
        # Cache project data
        await cache_service.cache_project_data(project["id"], project)
        
        return ProjectResponse(**project)
    
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )

@router.get("/", response_model=List[ProjectResponse])
async def get_user_projects(current_user: dict = Depends(get_current_user)):
    """Get all projects for the current user"""
    projects = await project_service.get_user_projects(current_user["id"])
    
    # Cache each project
    for project in projects:
        await cache_service.cache_project_data(project["id"], project)
    
    return [ProjectResponse(**project) for project in projects]

@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(
    project_id: str,
    current_user: dict = Depends(get_current_user)
):
    """Get a specific project"""
    # Check cache first
    cached_project = await cache_service.get_project_data(project_id)
    if cached_project and cached_project["user_id"] == current_user["id"]:
        return ProjectResponse(**cached_project)
    
    project = await project_service.get_project_by_id(project_id, current_user["id"])
    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found"
        )
    
    # Cache project data
    await cache_service.cache_project_data(project_id, project)
    
    return ProjectResponse(**project)

@router.put("/{project_id}", response_model=ProjectResponse)
async def update_project(
    project_id: str,
    updates: ProjectUpdate,
    current_user: dict = Depends(get_current_user)
):
    """Update a project"""
    update_data = {k: v for k, v in updates.dict().items() if v is not None}
    
    if not update_data:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No update data provided"
        )
    
    project = await project_service.update_project(
        project_id=project_id,
        user_id=current_user["id"],
        updates=update_data
    )
    
    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found"
        )
    
    # Update cache
    await cache_service.cache_project_data(project_id, project)
    
    return ProjectResponse(**project)

@router.delete("/{project_id}")
async def delete_project(
    project_id: str,
    current_user: dict = Depends(get_current_user)
):
    """Delete a project (soft delete)"""
    success = await project_service.delete_project(project_id, current_user["id"])
    
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found"
        )
    
    # Invalidate cache
    await cache_service.invalidate_project_cache(project_id)
    
    return {"message": "Project deleted successfully"}

File: backend-js/src/routes/projects.js
const express = require('express');
const { db } = require('../config/database');
const { cacheService } = require('../utils/cache');
const { authenticateToken } = require('./auth');

const router = express.Router();

// Create new project
router.post('/', authenticateToken, async (req, res) => {
  try {
    const { name, description, project_type = 'web' } = req.body;

    if (!name) {
      return res.status(400).json({ error: 'Project name is required' });
    }

    const projectData = {
      user_id: req.user.id,
      name,
      description: description || null,
      project_type,
      settings: {},
      is_active: true,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString()
    };

    const project = await db.createProject(projectData);

    // Cache project data
    await cacheService.set(`project:${project.id}`, project, 3600);

    res.status(201).json(project);
  } catch (error) {
    console.error('Create project error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get all user projects
router.get('/', authenticateToken, async (req, res) => {
  try {
    const projects = await db.getUserProjects(req.user.id);

    // Cache each project
    for (const project of projects) {
      await cacheService.set(`project:${project.id}`, project, 3600);
    }

    res.json(projects);
  } catch (error) {
    console.error('Get projects error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get specific project
router.get('/:projectId', authenticateToken, async (req, res) => {
  try {
    const { projectId } = req.params;

    // Check cache first
    const cachedProject = await cacheService.get(`project:${projectId}`);
    if (cachedProject && cachedProject.user_id === req.user.id) {
      return res.json(cachedProject);
    }

    const project = await db.getProjectById(projectId, req.user.id);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    // Cache project data
    await cacheService.set(`project:${projectId}`, project, 3600);

    res.json(project);
  } catch (error) {
    console.error('Get project error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Update project
router.put('/:projectId', authenticateToken, async (req, res) => {
  try {
    const { projectId } = req.params;
    const updates = req.body;

    // Remove undefined values
    const cleanUpdates = Object.fromEntries(
      Object.entries(updates).filter(([_, value]) => value !== undefined)
    );

    if (Object.keys(cleanUpdates).length === 0) {
      return res.status(400).json({ error: 'No update data provided' });
    }

    cleanUpdates.updated_at = new Date().toISOString();

    const project = await db.updateProject(projectId, req.user.id, cleanUpdates);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    // Update cache
    await cacheService.set(`project:${projectId}`, project, 3600);

    res.json(project);
  } catch (error) {
    console.error('Update project error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Delete project (soft delete)
router.delete('/:projectId', authenticateToken, async (req, res) => {
  try {
    const { projectId } = req.params;

    const result = await db.deleteProject(projectId, req.user.id);
    if (!result) {
      return res.status(404).json({ error: 'Project not found' });
    }

    // Invalidate cache
    await cacheService.delete(`project:${projectId}`);

    res.json({ message: 'Project deleted successfully' });
  } catch (error) {
    console.error('Delete project error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get project wireframes
router.get('/:projectId/wireframes', authenticateToken, async (req, res) => {
  try {
    const { projectId } = req.params;

    // Verify project ownership
    const project = await db.getProjectById(projectId, req.user.id);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    const wireframes = await db.getProjectWireframes(projectId);
    res.json(wireframes);
  } catch (error) {
    console.error('Get wireframes error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Create wireframe for project
router.post('/:projectId/wireframes', authenticateToken, async (req, res) => {
  try {
    const { projectId } = req.params;
    const { name, description, content_structure, design_system } = req.body;

    // Verify project ownership
    const project = await db.getProjectById(projectId, req.user.id);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }

    const wireframeData = {
      project_id: projectId,
      name: name || 'Untitled Wireframe',
      description: description || null,
      content_structure: content_structure || {},
      design_system: design_system || {},
      version_number: 1,
      is_active: true,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString()
    };

    const wireframe = await db.createWireframe(wireframeData);
    res.status(201).json(wireframe);
  } catch (error) {
    console.error('Create wireframe error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;
